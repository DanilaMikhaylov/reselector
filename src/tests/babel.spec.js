import React from 'react'
import { transformFileSync } from '@babel/core'
import { mount } from 'enzyme'

import {
  ClassComponent,
  ArrowFunctionalComponent,
  FunctionalComponent,
  ComposedComponent,
  ComposedComponent2,
  ComponentWithAttrs,
  SpreadPropsComponent,
} from './App'
import defaultNamedClassComponent from './App/defaultNamedClassExport'
import defaultDeclaredNamedClassComponent from './App/defaultDeclaredNamedClassExport'
import wrappedElementsComponent from './App/wrappedElements'
import { select } from '../'

describe('babel plugin', () => {
  describe('transform', () => {
    let config = {}

    beforeEach(() => {
      config = {}

      jest.doMock('../config', () => new Proxy(jest.requireActual('../config'), {
        get: (target, value) => (value in config ? config[value] : target[value]),
      }))
    })

    it('should transform source code well for development', () => {
      const { code } = transformFileSync(require.resolve('./App'))

      expect(code).toMatchSnapshot()
    })

    it('should transform source code depends on the env option', () => {
      config.env = true

      const { code } = transformFileSync(require.resolve('./App'))

      expect(code).toMatchSnapshot()
    })

    it('should transform default anonymous function', () => {
      const { code } = transformFileSync(require.resolve('./App/defaultAnonymousFunctionExport'))

      expect(code).toMatchSnapshot()
    })

    it('should transform default named function', () => {
      const { code } = transformFileSync(require.resolve('./App/defaultNamedFunctionExport'))

      expect(code).toMatchSnapshot()
    })

    it('should transform default named class', () => {
      const { code } = transformFileSync(require.resolve('./App/defaultNamedClassExport'))

      expect(code).toMatchSnapshot()
    })

    it('should transform declared named class in default', () => {
      const { code } = transformFileSync(require.resolve('./App/defaultDeclaredNamedClassExport'))

      expect(code).toMatchSnapshot()
    })

    it('should transform conditionals functions', () => {
      const { code } = transformFileSync(require.resolve('./App/conditionalRenderComponents'))

      expect(code).toMatchSnapshot()
    })
  })

  it('should find all the types of Components', () => {
    const components = [
      ClassComponent,
      ArrowFunctionalComponent,
      FunctionalComponent,
      ComposedComponent,
      ComponentWithAttrs,
      defaultNamedClassComponent,
      defaultDeclaredNamedClassComponent,
      SpreadPropsComponent,
      wrappedElementsComponent,
    ]

    components.forEach((Component) => {
      const wrapper = mount(<div><Component /></div>)

      const selector = select`${Component}`

      expect(wrapper).toMatchSnapshot()
      expect(wrapper.find(selector).hostNodes().length).toBe(1)
    })
  })

  it('should save Component`s root attrs', () => {
    const wrapper = mount(<ComponentWithAttrs />)
    const selector = select`${ComponentWithAttrs}`

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find(selector).getDOMNode().dataset).toMatchSnapshot()
  })

  it('should find nested Component', () => {
    const wrapper = mount((
      <ClassComponent>
        <ArrowFunctionalComponent>
          <FunctionalComponent>
            <ComposedComponent />
          </FunctionalComponent>
        </ArrowFunctionalComponent>
      </ClassComponent>
    ))

    const selector = select`
      ${ClassComponent}
      ${ArrowFunctionalComponent}
      ${FunctionalComponent}
      ${ComposedComponent}
    `

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find(selector).hostNodes().length).toBe(1)
  })

  it('should pass props through the React.Fragment', () => {
    const { code } = transformFileSync(require.resolve('./App/reactFragment'))

    expect(code).toMatchSnapshot()

    const Fragments = require('./App/reactFragment')

    Object.values(Fragments).forEach((Component) => {
      const wrapper = mount(<Component />)

      const selector = select`${Component}`

      expect(wrapper).toMatchSnapshot()
      expect(wrapper.find(selector).hostNodes().length).toBe(1)
    })
  })

  it('should find wrapped component', () => {
    const text = 'test text'
    const wrapper = mount(<ComposedComponent2>{text}</ComposedComponent2>)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find(select`${ComposedComponent2}`).hostNodes().text()).toBe(text)
  })
})
