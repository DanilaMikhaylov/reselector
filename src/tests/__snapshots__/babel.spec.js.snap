// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babel plugin should find all the types of Components 1`] = `
<div>
  <ClassComponent>
    <div
      data-a8f769be={true}
    />
  </ClassComponent>
</div>
`;

exports[`babel plugin should find all the types of Components 2`] = `
<div>
  <ArrowFunctionalComponent>
    <div
      data-f49764c6={true}
    />
  </ArrowFunctionalComponent>
</div>
`;

exports[`babel plugin should find all the types of Components 3`] = `
<div>
  <FunctionalComponent>
    <p
      data-b5d7faff={true}
    />
  </FunctionalComponent>
</div>
`;

exports[`babel plugin should find all the types of Components 4`] = `
<div>
  <ComposedComponent>
    <Button
      data-8e388bfc={true}
    >
      <button
        data-8e388bfc={true}
      />
    </Button>
  </ComposedComponent>
</div>
`;

exports[`babel plugin should find all the types of Components 5`] = `
<div>
  <ComponentWithAttrs>
    <div
      data-889909b2={true}
      data-test="testme"
    />
  </ComponentWithAttrs>
</div>
`;

exports[`babel plugin should find nested Component 1`] = `
<ClassComponent>
  <div
    data-a8f769be={true}
  >
    <ArrowFunctionalComponent>
      <div
        data-f49764c6={true}
      >
        <FunctionalComponent>
          <p
            data-b5d7faff={true}
          >
            <ComposedComponent>
              <Button
                data-8e388bfc={true}
              >
                <button
                  data-8e388bfc={true}
                />
              </Button>
            </ComposedComponent>
          </p>
        </FunctionalComponent>
      </div>
    </ArrowFunctionalComponent>
  </div>
</ClassComponent>
`;

exports[`babel plugin should save Component\`s root attrs 1`] = `
<ComponentWithAttrs>
  <div
    data-889909b2={true}
    data-test="testme"
  />
</ComponentWithAttrs>
`;

exports[`babel plugin should save Component\`s root attrs 2`] = `
DOMStringMap {
  "889909b2": "true",
  "test": "testme",
}
`;

exports[`babel plugin transform should transform source code depends on the env option 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.FunctionalComponent = FunctionalComponent;
exports.default = exports.ComponentWithAttrs = exports.ComposedComponent = exports.ArrowFunctionalComponent = exports.ClassComponent = void 0;

var _react = _interopRequireDefault(require(\\"react\\"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable react/prefer-stateless-function */
const Button = props => _react.default.createElement('button', props, props.children);
/* eslint-disable no-unused-vars */


const UnexportComponent = ({
  children
}) => _react.default.createElement(\\"div\\", {
  \\"data-c7b7156f\\": process.env.RESELECTOR === \\"true\\"
}, children);
/* eslint-enable no-unused-vars */


UnexportComponent.testId = \\"c7b7156f\\";

class ClassComponent extends _react.default.Component {
  render() {
    return _react.default.createElement(\\"div\\", {
      \\"data-a8f769be\\": process.env.RESELECTOR === \\"true\\"
    }, this.props.children);
  }

}

exports.ClassComponent = ClassComponent;
ClassComponent.testId = \\"a8f769be\\";

const ArrowFunctionalComponent = ({
  children
}) => _react.default.createElement(\\"div\\", {
  \\"data-f49764c6\\": process.env.RESELECTOR === \\"true\\"
}, children);

exports.ArrowFunctionalComponent = ArrowFunctionalComponent;
ArrowFunctionalComponent.testId = \\"f49764c6\\";

function FunctionalComponent({
  children
}) {
  return _react.default.createElement(\\"p\\", {
    \\"data-b5d7faff\\": process.env.RESELECTOR === \\"true\\"
  }, children);
}

FunctionalComponent.testId = \\"b5d7faff\\";

const ComposedComponent = ({
  children
}) => _react.default.createElement(Button, {
  \\"data-8e388bfc\\": process.env.RESELECTOR === \\"true\\"
}, children);

exports.ComposedComponent = ComposedComponent;
ComposedComponent.testId = \\"8e388bfc\\";

const ComponentWithAttrs = ({
  children
}) => _react.default.createElement(\\"div\\", {
  \\"data-test\\": \\"testme\\",
  \\"data-889909b2\\": process.env.RESELECTOR === \\"true\\"
}, children);

exports.ComponentWithAttrs = ComponentWithAttrs;
ComponentWithAttrs.testId = \\"889909b2\\";

var _default = ({
  children
}) => _react.default.createElement(\\"div\\", {
  \\"data-5f9b51d4\\": process.env.RESELECTOR === \\"true\\"
}, children);

exports.default = _default;

if (module.exports) {
  module.exports.default.testId = \\"5f9b51d4\\";
}"
`;

exports[`babel plugin transform should transform source code well for development 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.FunctionalComponent = FunctionalComponent;
exports.default = exports.ComponentWithAttrs = exports.ComposedComponent = exports.ArrowFunctionalComponent = exports.ClassComponent = void 0;

var _react = _interopRequireDefault(require(\\"react\\"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable react/prefer-stateless-function */
const Button = props => _react.default.createElement('button', props, props.children);
/* eslint-disable no-unused-vars */


const UnexportComponent = ({
  children
}) => _react.default.createElement(\\"div\\", {
  \\"data-c7b7156f\\": true
}, children);
/* eslint-enable no-unused-vars */


UnexportComponent.testId = \\"c7b7156f\\";

class ClassComponent extends _react.default.Component {
  render() {
    return _react.default.createElement(\\"div\\", {
      \\"data-a8f769be\\": true
    }, this.props.children);
  }

}

exports.ClassComponent = ClassComponent;
ClassComponent.testId = \\"a8f769be\\";

const ArrowFunctionalComponent = ({
  children
}) => _react.default.createElement(\\"div\\", {
  \\"data-f49764c6\\": true
}, children);

exports.ArrowFunctionalComponent = ArrowFunctionalComponent;
ArrowFunctionalComponent.testId = \\"f49764c6\\";

function FunctionalComponent({
  children
}) {
  return _react.default.createElement(\\"p\\", {
    \\"data-b5d7faff\\": true
  }, children);
}

FunctionalComponent.testId = \\"b5d7faff\\";

const ComposedComponent = ({
  children
}) => _react.default.createElement(Button, {
  \\"data-8e388bfc\\": true
}, children);

exports.ComposedComponent = ComposedComponent;
ComposedComponent.testId = \\"8e388bfc\\";

const ComponentWithAttrs = ({
  children
}) => _react.default.createElement(\\"div\\", {
  \\"data-test\\": \\"testme\\",
  \\"data-889909b2\\": true
}, children);

exports.ComponentWithAttrs = ComponentWithAttrs;
ComponentWithAttrs.testId = \\"889909b2\\";

var _default = ({
  children
}) => _react.default.createElement(\\"div\\", {
  \\"data-5f9b51d4\\": true
}, children);

exports.default = _default;

if (module.exports) {
  module.exports.default.testId = \\"5f9b51d4\\";
}"
`;
